### PR TITLE
aja: Capture and Output plugin for AJA Video Systems IO devices

### PR DESCRIPTION
**Pull request by paulh-aja**
# Description

## What is this?

This pull request consists of a capture and output plugin for AJA Video Systems SDI and HDMI I/O devices. Devices include those from AJA's Kona, Corvid and Io product lines, as well as the recently released T-TAP Pro SDI/HDMI output device. Development of the plugin was a collaborative effort between AJA and OBS developers.

Communication and data transfer between OBS and AJA devices is handled via the open-source AJA NTV2 SDK (MIT license), freely available for download from Github: https://github.com/aja-video/ntv2

Usage of the plugin requires installation of the AJA device driver. See device driver installation instructions in the "AJA Device Driver Installation" section of this document.

## Code Directories

The plugin and UI code lives in the following directories under `obs-studio`:

1. `plugins/aja` - Code for the actual capture/output functionality
1. `plugins/aja/sdk` - The AJA NTV2 SDK code
1. `UI/frontend-plugins/aja-output-ui` - Code for the AJA I/O/ Device Output UI pane

_NOTE: The most recent revision of the AJA NTV2 SDK is included in the AJA plugin sources in this PR. Building OBS from this PR will build both the AJA plugin and SDK sources. However, @jp9000 has indicated a preference for placing a pre-built ajantv2.lib into the OBS pre-built dependencies packages. --@paulh-aja_

### Adding an AJA I/O Device Capture source
1. Connect an AJA capture device to the system and install the AJA device driver.
1. Connect a valid HDMI or SDI signal to the device.
1. Open OBS.
1. Click on the plus (+) button in the Sources pane.
1. Click "AJA I/O Device Capture" in the Sources list.
aja_io_capture_add_source
1. Name the new AJA I/O Device Capture source.
aja_io_capture_create_add_existing
1. Select the desired AJA I/O Capture device from the device list.
1. Set Video Format and Pixel Format to "Auto".
1. Select the SDI or HDMI spigot on the device that your signal is connected to.
aja_io_capture_props_sdi1

NOTE: If the signal isn't properly auto-detected with Video and Pixel Formats set to "Auto" you can try manually setting the Video and/or Pixel Formats if you know what they are. An additional selection called "SDI 4K Transport" will appear when a UHD/4K video format is manually selected, allowing the user to select between 4K Squares (Square Division) and 2SI (Two Sample Interleave). Changing these options manually may help detect a signal when "Auto" fails to recognize the input signal.

### Starting the AJA I/O Device Output plugin (Program Output)
1. Connect an AJA capture device to the system and install the AJA device driver.
1. Connect a valid HDMI or SDI signal to the device.
1. Open OBS.
1. Select  the "Tools" -> "AJA I/O Device Output" menu item.
aja_output_tools_menu
1. Under "Program Output" select the AJA device to send the output signal through.
aja_output_props_default
1. Select the desired SDI/HDMI output spigot.
1. Select the desired Video and Pixel Format.
1. Click the Start button.

NOTE: To output the OBS Preview to an AJA device enable Studio Mode in OBS and follow the steps above for Program Output, substituting "Program Output" with the "Preview Output" section of the AJA I/O Device Output UI. Enabling _BOTH Program and Preview outputs at the same time_ will result in bad video outputs and is not yet supported (see Known Issues section).

## AJA Device Driver Installation

Follow one of the methods below to install the AJA device driver for your platform.

### Method 1 - AJA Software Installer (Windows/macOS/Ubuntu Linux)

_Pre-built_ driver packages for Windows, macOS and Ubuntu Linux are provided as a component of the AJA Software Installer. Navigate to the AJA Support page and download the latest AJA Software Installer for your platform:

https://www.aja.com/support

NOTE: An _open-source_ device driver for _Windows and macOS_ is not yet available at this time.

### Method 2 - Build From Sources (Linux-only at this time)

The _Linux verison_ of the AJA device driver (kernel module) can be built from the open-source (MIT license) AJA NTV2 SDK sources, found here: 

https://github.com/aja-video/ntv2

1. Clone the ntv2 repository.
1. `cd` into the `ntv2/ajadriver/linux` directory
1. Run `make`
1. `cd` into the `ntv2/bin` directory and run `sudo sh load_ajantv2` to install the AJA kernel module.

NOTE: To uninstall the Linux kernel module, `cd` into the `ntv2/bin` directory and run `sudo sh unload_ajantv2`.

# Motivation and Context

The motivation for the creation of a native OBS capture and output plugin for AJA devices is two-fold:

1. To unify AJA device support in OBS under a single native plugin, using a common code base for all three major platforms (Windows, macOS and Ubuntu Linux). This plugin utilizes the open-source AJA NTV2 device SDK (MIT license) rather than the usual platform-specific media APIs: DirectShow on Windows, CoreMediaIO on macOS and Video4Linux on Ubuntu Linux.
1. To serve as an initial test-bed for the recent open-source release of the AJA NTV2 device SDK under the MIT license.

We hope that the developer and user communities of OBS and AJA devices will find this plugin a welcome addition to the OBS ecosystem. While bug reports and contributions from the community are welcome and appreciated, AJA will continue to develop and support this new plugin.

# How Has This Been Tested?

The AJA OBS plugin was primarily developed under Windows 10, with additional development and testing performed on macOS Catalina and Big Sur, as well as Ubuntu Linux 18.04 and 20.04.

The primary Windows PCs used were a custom Intel i7 8700K/NVIDIA 1070ti desktop PC, two separate HP Z4 workstations, and a 2020 Dell XPS15 laptop.

macOS testing was performed on a 2014 Macbook Pro and two separate 2019 Mac Pro towers.

Limited Linux testing under Ubuntu 18.04 was performed on the aforementioned custom Intel i7 PC.

The following AJA capture/output devices were used during development and testing:

**PCI-E devices**
- Kona 4
- Kona 5
- Kona HDMI
- Corvid44

**Thunderbolt 2 devices**
- Io 4K

**Thunderbolt 3 devices**
- Io 4K Plus
- TTap Pro

Most devices in AJA Kona, Corvid and Io product lines _should_ work with the plugin but not all device, firmware and format combinations have been tested.

NOTE: The original Thunderbolt AJA T-TAP device is NOT supported by this plugin as it does not use the AJA NTV2 driver. The newer T-TAP Pro is supported however.

### Known issues and TODOs (ordered by severity)
1. Fix AV sync in the AJA I/O Output plugin with certain OBS/device frame rate combinations.
1. Enabling AJA Output for both Program and Preview simultaneously results in choppy video.
1. Improve frame tearing in the AJA I/O Device Capture plugin with certain OBS/device frame rate combinations.
1. Saved Capture plugin instances not auto-detecting signal properly if physical SDI/HDMI signal is not connected on re-launch of OBS, and then subsequently reconnected. Must twiddle the Capture properties to refresh the signal.
1. Output plugin colors and brightness look washed out. Colorspace conversion may not be applied correctly?
1. Need perform a thorough test matrix of video/pixel/SDI/HDMI transport settings and fill in any missing signal routing combinations. In general 4K and RGB need more testing.
1. Optimize framebuffer memory placement on device and coordinate between capture and output plugins.
1. Add support for Analog I/O capture and output on supported devices.
1. Certain combinations of video format with RGB pixel format needs additional testing.
1. Add support for selecting an alternate audio input source on supported devices (i.e. Microphone Input on AJA io4K+, analog audio, AES audio, etc). Capture and output currently default to using up to 8 channels of embedded SDI/HDMI audio.
1. All translation strings are placeholders and default to English. Translation strings needed for additional language support.
1. 8K video not yet implemented.
1. Add ability to display static text strings in OBS' generated UI system so that status or warning messages can be relayed to the user.

# Types of changes

- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

_Development Credits_
- @ddrboxman
- @paulh-aja

### CODE REVIEW

File: plugins/aja/update-locales.py
Comment (RytoEX) [inline plugins/aja/update-locales.py:14]: What purpose does this file serve?

**Review by RytoEX — CHANGES_REQUESTED** [review]
I've done an initial visual review of the code.  Many, if not all, of my review items can be considered to be a request to correct something throughout the PR.  I didn't highlight every specific instance because every such comment becomes an individual PR comment, which inflates the comment count and makes reading the thread difficult.

Ideally, I'd prefer if the SDK was not introduced in-tree into this repository.  It is a huge increase in number of files with no way to opt out of it while cloning the obs-studio repository.  If someone needs to check out the SDK to build it, they can do that separately.  If we need to include checkouts of the SDK in our CI/build process and/or prepare some prebuilt artifacts, then we can do that instead.

If the SDK absolutely must be included in-tree in the obs-studio repository, then at minimum, I'd prefer that the addition of the SDK is a separate commit.  Though again, ideally we'd pull in the SDK separately as a build-step if necessary.

Please remove all changes to non-"en-US" locales.  We track all translation work on Crowdin.
https://obsproject.com/forum/threads/how-to-contribute-translations-for-obs.16327/

Please include a CMake option to ENABLE/DISABLE building of the aja plugin.  This plugin increased build times on CI noticeably, and I would not want users who are not testing this plugin or AJA devices to have to build it locally.  Something like "ENABLE_AJA" in `plugins/aja/CMakeLists.txt` that returns early if false is fine.

Please split the commits appropriately for adding the "aja" plugin and the "aja-output-ui" frontend plugin.

There are several places where you disable and re-enable clang-format.  While I think that we should not enforce clang-format in certain areas, like third-party libraries or SDKs that we have in-tree, I do think that we should avoid disabling clang-format in other places such as normal C/C++ files unless there is an extremely good reason to do so.  I can see in some places this is done to maintain table-like formatting for arrays, and that is probably acceptable.  In other places, it seems like it's just used to avoid the line length limit, which I don't think is a great justification.

That's all that I can think of for now.

**Comment by kkartaltepe** [comment]
One small nit, please add an option to allow disabling building of this plugin.

Though it is self contained at the moment, once the sdk is an external dependency there will be a need to allow users to compile without this plugin. Its awesome to see more companies open sourcing their userspace and even linux drivers, but on debian/ubuntu with long lived releases it will take time before this plugin could be a hard dependency by everyone building on every distro.

```cmake
option(ENABLE_AJA "Enable AJA plugin support" ON)
if(NOT ENABLE_AJA)
    message(STATUS "AJA plugin disabled")
    return()
endif()
```

Should be good enough.

File: UI/frontend-plugins/aja-output-ui/CMakeLists.txt
Comment (kkartaltepe) [reply UI/frontend-plugins/aja-output-ui/CMakeLists.txt:96]: I dont see `QX11Info` being included so this can probably just be removed?

**Comment by paulh-aja** [comment]
> One small nit, please add an option to allow disabling building of this plugin.
> 
> Though it is self contained at the moment, once the sdk is an external dependency there will be a need to allow users to compile without this plugin. Its awesome to see more companies open sourcing their userspace and even linux drivers, but on debian/ubuntu with long lived releases it will take time before that work will be available and we can make this plugin could be a hard dependency by everyone building on every distro.
> 
> ```cmake
> option(ENABLE_AJA "Enable AJA plugin support" ON)
> if(NOT ENABLE_AJA)
>     message(STATUS "AJA plugin disabled")
>     return()
> endif()
> ```
> 
> Should be good enough, in `aja/CMakeLists.txt`
> 
> --- edit
> Looks like rytoex beat me too it!

No problem, I'll take care of it.

File: UI/frontend-plugins/aja-output-ui/CMakeLists.txt
Comment (paulh-aja) [reply UI/frontend-plugins/aja-output-ui/CMakeLists.txt:96]: This AJA Output UI code started with the Decklink Output UI code as a base. I'll take a look at changes to the Decklink code where the Qt dependencies have changed since work on the AJA plugin was started.

File: UI/frontend-plugins/aja-output-ui/forms/output.ui
Comment (paulh-aja) [reply UI/frontend-plugins/aja-output-ui/forms/output.ui:130]: OK, this can be removed for the time being. This was originally copied from the Decklink Output UI code.

File: plugins/aja/update-locales.py
Comment (paulh-aja) [reply plugins/aja/update-locales.py:14]: That shouldn't have gotten in there. I'll remove it. I used it during testing to generate the additional locales files from the source of the en-US file.

File: plugins/aja/aja-card-manager.cpp
Comment (paulh-aja) [reply plugins/aja/aja-card-manager.cpp:309]: Should I move the comments so that they appear before the entire if-else block?
For example:
```
	/* Handle checking special case outputs before all other outputs.
	 * 1. HDMI Monitor uses framestore 4
	 * 2. SDI Monitor on Io 4K/Io 4K Plus, etc. uses framestore 4.
	 * 3. Everything else...
	 */
	if (aja::CardCanDoHDMIMonitorOutput(id) &&
	    io == IOSelection::HDMIMonitorOut) {
		NTV2Channel hdmiMonChannel = NTV2_CHANNEL4;
		return ChannelReady(hdmiMonChannel, owner);
	} else if (aja::CardCanDoSDIMonitorOutput(id) &&
		   io == IOSelection::SDI5) {
    ...
```

File: plugins/aja/aja-card-manager.cpp
Comment (paulh-aja) [reply plugins/aja/aja-card-manager.cpp:160]: OK, I will change it to "Unknown" here. I was mimicking a function in the AJA NTV2 SDK which returns "???" for unknown NTV2DeviceID strings.

File: plugins/aja/aja-card-manager.cpp
Comment (paulh-aja) [reply plugins/aja/aja-card-manager.cpp:270]: That was a typo. The comment should rather indicate that the `UNUSED_PARAMETER(id)` is a placeholder for device-specific configuration for the capture plugin, following the structure of the other functions where device-specific configuration was implemented for the output plugin (`AcquireOutputSelection` and `ReleaseOutputSelection`).

Comment (RytoEX) [reply plugins/aja/aja-card-manager.cpp:270]: I think it would be apparent to most people who might modify these functions that if they added code to use the parameter in question (`id`) that they should remove the call to the `UNUSED_PARAMETER` macro.

File: plugins/aja/aja-card-manager.cpp
Comment (RytoEX) [reply plugins/aja/aja-card-manager.cpp:309]: Either placing them before the entire statement, as you have demonstrated, or placing an appropriate comment within each block would be fine.  Alternatively, placing a detailed comment block before the entire statement and also placing short single line summary comments in each block would also be fine.

If the comments become too unwieldy or get in the way of reading the code, I'm sure someone else will mention it during the review process.

**Comment by paulh-aja** [comment]
> Please include a CMake option to ENABLE/DISABLE building of the aja plugin. This plugin increased build times on CI noticeably, and I would not want users who are not testing this plugin or AJA devices to have to build it locally. Something like "ENABLE_AJA" in `plugins/aja/CMakeLists.txt` that returns early if false is fine.

Is there any way I can view the CI builds where this PR is building?

**Comment by kkartaltepe** [comment]
> Is there any way I can view the CI builds where this PR is building?

https://github.com/obsproject/obs-studio/blob/master/.github/workflows/main.yml#L353

**Comment by kilinbox** [comment]
Screen Shot 2021-09-17 at 21 15 28

Thank you very much for your work.  
I used KonaHDMI to capture my PS5 video.  
I couldn't select YUV-10 for Pixcel Format, which caused problems with the color representation.

**Comment by paulh-aja** [comment]
> > Is there any way I can view the CI builds where this PR is building?
> 
> https://github.com/obsproject/obs-studio/blob/master/.github/workflows/main.yml#L353
> 
> --- edit
> Unless you mean the results those are in the checks you can review the logs.

I see it now. Thanks.

**Comment by paulh-aja** [comment]
> Screen Shot 2021-09-17 at 21 15 28
> 
> Thank you very much for your work.
> I used KonaHDMI to capture my PS5 video.
> I couldn't select YUV-10 for Pixcel Format, which caused problems with the color representation.

I haven't implemented a true 10-bit video path in OBS yet but the ability to capture 10-bit YUV with an YUV-8 framebuffer should work. I'll have to take a look at this w/ Kona HDMI.

**Comment by kilinbox** [comment]
> If your signal is in fact sending UHDp59.94 RGB then that will not work in the current version of the plugin. I have not yet implemented the settings for capturing or outputting UHD/4K RGB "high-frame-rate" (i.e 50/60fps) video, although the Kona HDMI is capable of capturing one stream of UHD/4K RGB 60fps video. It is on my TODO list.
> 
Screen Shot 2021-09-18 at 8 32 36

Thanks. It turns out that my settings were wrong.
I was able to solve the problem by outputting as YUV video.

**Comment by kilinbox** [comment]
screen

https://twitter.com/kilinbox/status/1439040334332248066

Sorry again.
It looks like the channels are arranged differently than in the AJA Control Room. 
Is there something I need to set?

**Comment by paulh-aja** [comment]
> screen
> 
> https://twitter.com/kilinbox/status/1439040334332248066
> 
> Sorry again.
> It looks like the channels are arranged differently than in the AJA Control Room.
> Is there something I need to set?

I spoke with some of the other AJA engineers about this issue and there apparently is a known issue with audio channel re-ordering on the Kona HDMI specifically. I'll need to look at the AJA Control Room code to see how they handle it and apply the fix the OBS plugins. Stay tuned, and thanks for the report. :)

**Comment by paulh-aja** [comment]
This PR has been updated to use external deps for the AJA NTV2 SDK.

I've also added a CMake variable `DISABLE_AJA` to manually disable building of the AJA plugin.

Eventually the NTV2 SDK headers + static libs will go into the OBS pre-built dependencies. In the meantime, here are the instructions for building the plugin by manually placing the NTV2 headers and static libs in the OBS `${DepsPath}` location:

```
git clone git@github.com:paulh-aja/obs-studio.git
git checkout aja-plugin
```

Download the NTV2 16.1 static libs + headers: https://github.com/aja-video/ntv2/releases/download/16.1/ntv2_16_1_static_libs.zip

**Windows**
1. Unzip the OBS Windows pre-built dependencies (Dependencies2019.zip). This location is used by the OBS CMake build as `${DepsPath}`.
2. Unzip `ntv2_16_1_static_libs.zip` and copy the following directories into `${DepsPath}/include/ajalibraries`:
- `ajaanc`
- `ajabase`
- `ajantv2`
3. Copy the following static libs from the unzipped `ntv2_16_1_static_libs.zip` into `${DepsPath}/bin`
- `libs/Debug/windows/x64/ajantv2d.lib`
- `libs/Release/windows/x64/ajantv2.lib`
4. Run the OBS CMake build with `${DepsPath}` set to the location of the unzipped pre-built dependencies. The AJA plugin will build if the NTV2 16.1 headers and static libs were placed in the correct pre-built deps locations.
5. Build OBS from the generated CMake make files.

**macOS**
1. If it does not yet exist, create `/tmp/obsdeps/ajalibraries` and `/tmp/obsdeps/bin`.
2. Unzip `ntv2_16_1_static_libs.zip` and copy the following directories into `/tmp/obsdeps/include/ajalibraries`:
- `ajaanc`
- `ajabase`
- `ajantv2`
3. Copy the following static libs from the unzipped `ntv2_16_1_static_libs.zip` into `/tmp/obsdeps/bin`:
- `libs/Debug/macos/x64/libajantv2d.a`
- `libs/Release/macos/x64/libajantv2.a`
NOTE: If building OBS for native arm64, copy the static libs from the `arm64` directory instead of `x64`.
4. Run the OBS `CI/full-build-macos.sh` script to build OBS. This script already uses `/tmp/obsdeps` as the `${DepsPath}` CMake variable, and should pick up the NTV2 16.1 dependencies if they were copied to the correct sub-directories of `/tmp/obsdeps`.

**Linux**
1. Create a directory to use for pre-built dependencies. This location will be used as `${DepsPath}` in the OBS CMake build.
2. Unzip `ntv2_16_1_static_libs.zip` and copy the following headers directories into `${DepsPath}/include/ajalibraries`:
- `ajaanc`
- `ajabase`
- `ajantv2`
3. Copy the following static libs from `ntv2_16_1_static_libs.zip` into `${DepsPath}/bin`:
- `libs/Debug/linux/x64/libajantv2d.a`
- `libs/Release/linux/x64/libajantv2.a`
4. Run OBS CMake to generate make files, with the CMake variable `DepsPath` set to the pre-built dependencies location.
5. Run `make` with the generated OBS CMake Makefile.

**Comment by RytoEX** [comment]
Please rebase your work on top of the current git master instead of merging branches.

**Comment by paulh-aja** [comment]
> The source repository for ajalibraries does not have issues enabled, so I will ask here. Please make it possible to build the release artifacts (e.g. the zip file you released) via your cmake. Im not sure about how the layout for the current release was decided but personally as long as you include a pkg-config folder with all the right flags im happy.
> 
> This is just my thoughts as its unlikely your existing repository will be picked up as is by any linux distribution. Probably not required for this PR unless someone feels strongly.

I have some changes to the CMake scripts, as well as a Python script that runs on our internal CI system to generate the static libs for that release zip. I can push these changes to the NTV2 github repo, no problem. I just hadn't gotten to it yet as I've been re-working how we generate the open-source repo from our internal source tree.

We do plan to enable issues on Github once we've formulated a policy and process within our team for handling issue feedback.

**Comment by RytoEX** [comment]
> 
> 
> > > > Please rebase your work on top of the current git master instead of merging branches.
> > > 
> > > 
> > > Sorry, I'm still kind of a git newbie. I had made some changes in another branch `aja-plugin-extdeps`, and then rebased that on OBS `master` before merging it with `aja-plugin` branch.|
> > > What do I need to do in order to fix this?
> > 
> > 
> > In your local Git repo, if you have your fork (paulh-aja/obs-studio) set as a remote named "origin" and obsproject/obs-studio set as a separate remote named "upstream" (you can check your remotes by executing `git remote -v`), then the following should get you to a cleaner state:
> > ```
> > git checkout master
> > git fetch upstream
> > git pull --ff-only upstream master
> > git checkout aja-plugin
> > git rebase master
> > git push -f origin
> > ```
> > 
> > You could do the same for any branch of obs-studio that you're working on. This process works fine for me and for others I've walked through it. If you're concerned about messing up your local branch, you can always make a backup first.
> 
> @DDRBoxman walked me through it. I figured I'd probably have to do a force push but I wasn't sure if there was another way around it. Does the `aja-plugin` branch look OK now with regards to your previous comment re: rebase?

Yes.  While before this PR consisted of over 400 files with over 400K lines and 90+ commits, this is now a much more manageable 42 files, about 10K lines, and 10 commits.  It is likely that at some point we'll ask that all the commits be squashed into a single commit, but that can be done at any point before merging.

**Comment by paulh-aja** [comment]
> > > > > Please rebase your work on top of the current git master instead of merging branches.
> > > > 
> > > > 
> > > > Sorry, I'm still kind of a git newbie. I had made some changes in another branch `aja-plugin-extdeps`, and then rebased that on OBS `master` before merging it with `aja-plugin` branch.|
> > > > What do I need to do in order to fix this?
> > > 
> > > 
> > > In your local Git repo, if you have your fork (paulh-aja/obs-studio) set as a remote named "origin" and obsproject/obs-studio set as a separate remote named "upstream" (you can check your remotes by executing `git remote -v`), then the following should get you to a cleaner state:
> > > ```
> > > git checkout master
> > > git fetch upstream
> > > git pull --ff-only upstream master
> > > git checkout aja-plugin
> > > git rebase master
> > > git push -f origin
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > You could do the same for any branch of obs-studio that you're working on. This process works fine for me and for others I've walked through it. If you're concerned about messing up your local branch, you can always make a backup first.
> > 
> > 
> > @DDRBoxman walked me through it. I figured I'd probably have to do a force push but I wasn't sure if there was another way around it. Does the `aja-plugin` branch look OK now with regards to your previous comment re: rebase?
> 
> Yes. While before this PR consisted of over 400 files with over 400K lines and 90+ commits, this is now a much more manageable 42 files, about 10K lines, and 10 commits. It is likely that at some point we'll ask that all the commits be squashed into a single commit, but that can be done at any point before merging.

@DDRBoxman told me that it would be OK to keep adding commits to my `aja-plugin` branch, as they would eventually be squashed when merged to OBS master.

**Comment by PatTheMav** [comment]
My vote is against including the SDK in `obs-deps`, but instead have CMake treat the SDK as an "ExternalProject" so it's downloaded and added as a library target if the plugin is not disabled via the cache setting mentioned above (moving forward I'd require this for any plugin/functionality that requires external SDKs):

https://cmake.org/cmake/help/latest/module/ExternalProject.html

**Comment by kkartaltepe** [comment]
> The structure of the Release zip was somewhat arbitrary. Our initial goal was first and foremost to build a first-class OBS capture and output plugin using our SDK, which required releasing our AJA NTV2 code under an open-source license. We do want to continue to massage the open-source SDK into a form that would meet the expectations of the wider open-source and Linux communities though.
> 
> Would you mind pointing me toward an example of a particularly well-structured lib that is included into Linux distros? I'll work on getting our NTV2 repo structure more in line with expected norms in Linux libraries.

Here (https://gist.github.com/kkartaltepe/4b8eaf13aff3c13059a088d1ee09e249) is an example of what could be added to make your existing cmake configuration add install targets and install the appropriate libs/static libs/headers. I just use the default flat `lib` folder and `include/<libname>/<headers>` structure. Havnt not looked at the actual `#include`'s in the headers I'm not sure such a layout is workable or if you need all the subdirectories constructed in the release zip, but if they are required my suggestion would have a structure like `include/libname/sub/dirs/header.h`.

Such install targets do not make a multi-platform+debug+release artifact work if you are really set on that you would probably need keep a similar /lib setup to what you already have but include a pkg-config for each platform along with its binary.

The pkg-config part is an example of what a pkg-config setup might look like. But i'm not sure I understand the dependency situation for your libs to say it's actually a correct pkg-config.

For all of this maybe Pat has some better suggestions having spent a lot of time on modernizing our own cmake. Personally I find external projects more configuration and more fragile than they are worth (compared to installing the project into a prefix and using a standard Find*.cmake) in my limited experience with them, but I defer to Pat.

**Comment by paulh-aja** [comment]
> It kinda depends on how Jim wants to integrate this - if it should become a fixed part of OBS, then shipping the library with obs-deps might make more sense (vs shipping the source code like we do for `ftl` which is deprecated-ish), then we'd need this library to be available via apt, and ready to compile for Windows (32bit and 64bit) and macOS (x86_64 and arm64, preferably universal) and use an appropriate Finder.
> 
> If it's an optional feature, then it should be a standalone plugin with its own installer that e.g. fetches the library (like our NDI plugin).

When we first started developing this plugin w/ @DDRBoxman , the original plan was supposed to be that the ajantv2 would be used as a shared library. It quickly became apparent that we lacked the facilities in our code for this to actually work, particularly on Windows due to search path issues, not to mention the name mangling problem. Up until this point the ajantv2 library had only ever been used by OEM integrators, directly in their products, as a static lib or compiled in directly.

Building infrastructure in our code to load classes and functions through a shared lib on the system looked like it was going to take a lot more work than we'd originally planned for this project. We'd originally hoped to have the plugin completed for integration into the last major OBS release (27.0).

At a minimum we'd need a way to find an ajantv2 shared lib on the system in a number of search paths, and factory classes to instantiate various NTV2 C++ objects. We'd also probably need to do this in such a way that could handle versioning of the NTV2 lib. The Decklink software has a robust system to handle this via a kind of COM interface, and IDL. Eventually we may get our code into the state where it will be usable via shared libs but today it just isn't there.

Through discussions w/ @jp9000, @DDRBoxman and @dodgepong we decided it would be OK to put ajantv2 static libs and headers in the OBS pre-built deps package and add the appropriate "FindLibrary.cmake" finder script in OBS, which is now done in this PR. It is also entirely possible to build the static libs via the CMake in our NTV2 repository (https://github.com/aja-video/ntv2).

**Pull request merged**
